### PR TITLE
Add documentation for Prefect Integration

### LINKED ISSUE #404: Documentation for Prefect
Prefect is implemented but requires documentation. @aktech has the most knowledge on this

### PR DESCRIPTION
**Pull request by aktech**
Fixes #404 #661 

The fix for prefect issues is here: #667

### CODE REVIEW

File: docs/source/08_integrations/1_prefect.md [hunk 1]
```diff
@@ -0,0 +1,173 @@
+# Prefect
+
+Prefect is a workflow automation system. QHub integrates Prefect with a
+feature flag as follows:
+
+```yaml
+prefect: true
```
Comment (kcpevey) [inline docs/source/08_integrations/1_prefect.md:7]: at what level in the config yaml does this belong? went to another project to see that its at the zero indent level.

File: docs/source/08_integrations/1_prefect.md [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+Prefect Agents is a lightweight processes for orchestrating flow runs. Agents
+run inside a user's architecture, and are responsible for starting and monitoring
+flow runs.
+
+During operation the agent process queries the Prefect API for any scheduled flow
+runs, and allocates resources for them on their respective deployment platforms.
+
+When you create set the flag `prefect` to `true` in the `qhub-config.yml` prefect
+agen is deployed on the QHub's kubernetes cluster.
```
Comment (kcpevey) [inline docs/source/08_integrations/1_prefect.md:42]:
```suggestion
agent is deployed on the QHub's kubernetes cluster.
```

File: docs/source/08_integrations/1_prefect.md [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+You also need a token/API Key to communicate to Prefect Cloud, same can be generate
+via Prefect Dashboard.
+
+1. Create a free Prefect cloud account here: https://cloud.prefect.io/
+2. Create a Service Account and an API key for the same
+3. Create a project in the Prefect Cloud Dashboard. Alternatively from CLI:
+
+```
+prefect create project 'your-prefect-project-name
```
Comment (kcpevey) [inline docs/source/08_integrations/1_prefect.md:21]:
```suggestion
prefect create project 'your-prefect-project-name'
```

File: docs/source/08_integrations/1_prefect.md [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+feature flag as follows:
+
+```yaml
+prefect: true
+```
+
+There are a bunch of components in getting Prefect working for you, here
+is a brief description of them:
+
+You also need a token/API Key to communicate to Prefect Cloud, same can be generate
```
Comment (peytondmurray) [inline docs/source/08_integrations/1_prefect.md:13]: Would be useful to include some details on how to use the API key once you've generated it.

**Comment by viniciusdc** [comment]
@aktech @tylerpotts I think there is an error with prefect deployment. Folowing a slack question the was rised today, when @kcpevey tried to deploy QHub using 

```yaml
prefect: true
```
She got the folowing error message when executing: `qhub deploy --config qhub-config.yaml --disable-prompt --skip-remote-state-provision`

```bash
Problem encountered: 1 validation error for Main
prefect
  extra fields not permitted (type=value_error.extra)
```

Folowing that, I decided to make a clean local dpeloyment of QHub and add `prefect: true` to the `qhub-config.yml`, receving the folowing error when attempting to peform `qhub render`

```bash
Problem encountered: 1 validation error for Main
prefect
  extra fields not permitted (type=value_error.extra)
``` 

As the problem is the same, my thoght is that Pydantic is not recogninzing the new field for prefect, I manage to solve the problems adding
```python
    prefect: typing.Optional[bool]
```
to the `Main(Base)` class of `schema.py`.

**Comment by kcpevey** [comment]
We are missing documentation on how to connect the qhub prefect agent to the prefect cloud account.

**Comment by aktech** [comment]
> We are missing documentation on how to connect the qhub prefect agent to the prefect cloud account.

It's here: https://github.com/Quansight/qhub/pull/641/files#diff-135df044ba5e1176d73cc20cc673aba03c3b7e8339661932529ece0cce7ddceaR14

**Comment by kcpevey** [comment]
First, I'm confused by the instructions asking me to add an env var called `PREFECT_TOKEN` when the [prefect docs](https://docs.prefect.io/orchestration/tutorial/overview.html#authenticating-with-prefect-cloud) talk only mention `PREFECT__CLOUD__AUTH_TOKEN` and `PREFECT__CLOUD__AGENT__AUTH_TOKEN`.

Second, there is some additional confusion around `what qhub does` vs `what my client package will need to do` AND what is needed for both. My guess is that qhub is taking `PREFECT_TOKEN` and is probably renaming it into one of those other variables. If that's done, then does that mean my package won't need to set these? If so, I could get away with just a quick check to ensure it exists.

**Comment by kcpevey** [comment]
Also, secrets are very different in gitlab vs github. I'm working in gitlab and I'm hoping that by `add it CI secrets` you mean `add it as an env var`.

**Comment by iameskild** [comment]
I'm running into the similar problems. I've tried deploying a prefect-agent a few times and each results prefect-agent pod with "No agent API token provided" error. For references I have tried the following:
- Using `PREFECT_TOKEN` env var with a service account API key 
- Using `PREFECT__CLOUD__AGENT__AUTH_TOKEN` env var with a service account API key
- Manually entering the service account API key when prompted during the deployment (`Enter a value`)

Could it be that we need to pass the token when we start the (prefect) kubernetes agent here:
https://github.com/Quansight/qhub/blob/b530bc3019acf748034130b3a13519bd11263e15/qhub/template/%7B%7B%20cookiecutter.repo_directory%20%7D%7D/infrastructure/modules/kubernetes/services/prefect/chart/templates/prefect.yaml#L21

I can see further down that some env vars are being passed in but perhaps the agent needs the token at the time it starts?
Something like:
```
prefect agent start -t PREFECT__CLOUD__AUTH__AGENT_TOKEN=$PREFECT_TOKEN -e ...
```

File: docs/source/08_integrations/1_prefect.md (hunk 1 above)
Comment (aktech) [reply docs/source/08_integrations/1_prefect.md]: At zero/top level. Anything that's not indented with another top level key is assumed to be top level.

File: docs/source/08_integrations/1_prefect.md (hunk 4 above)
Comment (aktech) [reply docs/source/08_integrations/1_prefect.md]: added

**Comment by aktech** [comment]
The fix for prefect issues is here: https://github.com/Quansight/qhub/pull/667

**Comment by aktech** [comment]
I have added the documentation for the whole process, which belongs to QHub. Let me know if anything is not clear.

File: docs/source/08_integrations/1_prefect.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+1. Create a free Prefect cloud account here: https://cloud.prefect.io/
+2. Create a Service Account and an API key for the same and add this to the CI secrets as `TF_VAR_prefect_token`:
+   - In GitHub: Set it in Secrets (https://docs.github.com/en/actions/reference/encrypted-secrets#creating-encrypted-secrets-for-a-repository)
+   - In GitLab: Set it as Variables (https://docs.gitlab.com/ee/ci/variables/#gitlab-cicd-variables)
+3. Create a project in the Prefect Cloud Dashboard. Alternatively from CLI:
+
+The `TF_VAR_prefect_token` API key is set as `PREFECT__CLOUD__AGENT__AUTH_TOKEN`
+environment variable in the agent. It is used while deploying Prefect Agent so that
+it can connect to Prefect Cloud and query flows.
```
Comment (kcpevey) [inline docs/source/08_integrations/1_prefect.md:30]: this is great - thanks!

**Review by kcpevey — APPROVED** [review]
LGTM! Thanks for the clarification! :)

**Pull request merged**